Framework/Algorithms/src/FFTPreProcessing.cpp

```diff
@@ -0,0 +1,250 @@
+                           applyApodizationFunction(
+                               addPadding(inputWS->histogram(specNum), padding),
+                               decayConstant, apodizationFunction));
+    prog.report();
+    PARALLEL_END_INTERUPT_REGION
+  }
+  PARALLEL_CHECK_INTERUPT_REGION
+  setProperty("OutputWorkspace", outputWS);
+}
+
+typedef double (*fptr)(const double time, const double decayConstant);
+/**
+* Gets a pointer to the relevant
+* apodization function
+* @param method :: [input] The name of the chosen function
+* @returns :: pointer to the function
+*/
+fptr FFTPreProcessing::getApodizationFunction(const std::string method) {
+  if (method == "None") {
+    return none;
+  } else if (method == "Bartlett") {
+    return bartlett;
+  } else if (method == "Lorentz") {
+    return lorentz;
+  } else if (method == "Connes") {
+    return connes;
+  } else if (method == "Cosine") {
+    return cosine;
+  } else if (method == "Gaussian") {
+    return gaussian;
+  } else if (method == "Welch") {
+    return welch;
+  }
+  throw std::invalid_argument("The apodization function selected "+
+	  method +
+	  " is not a valid option");
+}
+/**
+* Applies the appodization function to the data.
+* @param histogram :: [input] Input histogram.
+* @param function :: [input] the apodization function
+* @param decayConstantn :: [input] the decay constant for apodization function
+* @returns :: Histogram of the apodized data
+*/
+HistogramData::Histogram FFTPreProcessing::applyApodizationFunction(
+    const HistogramData::Histogram &histogram, const double decayConstant,
+    fptr function) {
+  HistogramData::Histogram result(histogram);
+  result.points();
+  auto &xData = result.mutableX();
+  auto &yData = result.mutableY();
+  auto &eData = result.mutableE();
+
+  for (size_t i = 0; i < yData.size(); ++i) {
+    yData[i] *= function(xData[i], decayConstant);
+    eData[i] *= function(xData[i], decayConstant);
+  }
+  result.binEdges();
+  return result;
+}
+/**
+* Adds padding to the data. The padding is
+* an integer multiple of the original data set.
+* i.e. padding =0 => none
+* padding = 2 => 2/3 of the output will be zero.
+* @param histogram :: [input] Input histogram
+* @param padding :: [input] the amount of padding to add
+* @returns :: Histogram of the padded data
+*/
+HistogramData::Histogram
+FFTPreProcessing::addPadding(const HistogramData::Histogram &histogram,
+                             const int padding) {
+
+  HistogramData::Histogram result(histogram);
+  // make sure point data
+  result.points();
+  auto &xData = result.x();
+  auto &yData = result.y();
+  auto &eData = result.e();
+  // assume approx evenly spaced
+  double dx = xData[1] - xData[0];
+  std::vector<double> newXData;
+  std::vector<double> newYData;
+  std::vector<double> newEData;
+
+  if (getProperty("NegativePadding")) {
+	  double xValue = xData[0] - dx*yData.size()*padding;
+	  for (int j = 0; j < padding; j++) {
```

Comment:
Sorry, forgot to mention that when padding on both ends each end gets half of the one-sided padding. So when `padding == 1` the size of the data doubles in both cases.